### ISSUE TITLE
Change Headers for "Source image is not a valid file" from 500 to 404

### ISSUE THREAD

**Issue by Anatolie** [mosbth/cimage#127]
When doing testing I noticed that an invalid source name throws a 500 Internal Server Error header. Recommend having the script throw a 404 Not Found header instead, allowing redirects to a 404 page.

**Comment by mosbth** [comment]
Yes, seems like a good suggestion.

**Comment by Anatolie** [comment]
Proposed change for the errorPage() function:

```php
function errorPage($msg, $type=500)
{
    global $mode;

	switch ($type) {
		case 404:
			$header = "HTTP/1.0 404 Not Found";
		break;
		default:
			$header = "HTTP/1.0 500 Internal Server Error";
	}

    header($header);

    if ($mode == 'development') {
        die("[img.php] $msg");
    }

    error_log("[img.php] $msg");
    die($header);
}
```

And then simply change line 4035 (*referencing imgp.php*) to:

```php
or errorPage(
    'Source image is not a valid file, check the filename and that a
    matching file exists on the filesystem.'
, 404);
```

Switch in the error function will allow to add more error messages as well. Haven't had the chance to look through places to place different headers, but would suggest adding the following as well:

* 400 Bad Request
* 401 Unauthorized
* 403 Forbidden

Will push a commit when I get a chance to look through all errorPage() calls :smiley:

**Comment by mosbth** [comment]
Lets wait with 400, 401 and 403 until we found a use for them. There might
be some use, as somethings are forbidden, or can be forbidden by
configuration. But we also want to mirror how an ordinary <img> works. But
the idea is good, so lets see what you find..

Change img.php, and not img[d,p,s].php since those are all created as a
result from running the script bin/create-img-single.bash

Cheers,
Mikael

**Comment by Anatolie** [comment]
Pushed out a pull request with changes made to img.php and the other two files [p,d], but those can be ignored since you mentioned they're generated by a script.

Let me know what you think of the changes proposed, each error message has an associated error code now.

**Comment by Anatolie** [comment]
Good point about mirroring the <img> tags, I'll research that in more detail and make a pull request with any necessary changes. Thanks for the great script! :+1:

**Comment by mosbth** [comment]
Just to review and document what you propose.

These are the status code we could use (from your pull req #128).

| Status codes | 
|------------------|
| 400 Bad Request
| 401 Unauthorized
| 403 Forbidden
| 404 Not Found
| 500 Internal Server Error

Currently all errors generates a status code of 500.

Your suggestion is like this.

| Status code proposed | Message or what happened |
|:---------------:|-------------------------------------|
| 500 | All exceptions thrown by `CImage`.
| 500 | Extension gd is not loaded.
| 500 | Unknown mode: $mode (relates to development, production, strict mode)
| 401 | Password required and does not match or exists.
| 401 | Hotlinking/leeching not allowed when password missmatch.
| 403 | Hotlinking/leeching not allowed and referer is missing.
| 403 | Hotlinking/leeching not allowed by whitelist. Referer: $referer.
| 403 | Hotlinking/leeching not allowed.
| 400 | Must set src-attribute.
| 400 | Filename contains invalid characters.
| 404 | Source image is not a valid file, check the filename and that a matching file exists on the filesystem.
| 500 | Security constraint: Source image is not below the directory "image_path" as specified in the config file img_config.php.
| 400 | Width % not numeric.
| 400 | Width out of range.
| 400 | Height % out of range.
| 400 | Height out of range.
| 400 | Aspect ratio out of range
| 400 | Quality out of range
| 400 | Compress out of range
| 400 | Scale out of range
| 400 | RotateBefore out of range
| 400 | RotateBefore out of range
| 500 | Directory for alias is not writable.
| 500 | Filename for alias contains invalid characters. Do not add extension.
| 500 | Alias is not enabled in the config file or password not matching.

Lets take a moment and think about this, just to select the proper status codes and have a easy-to-grasp logic for this. 

*And* bring into the equation how the different modes should deal with this (development, production, strict), perhaps strict should only deliver 500 or 404 trying to mask/hide itself from being operated by cimage.

*And* if this should be configurable in the config-file, perhaps to override it all and do ones own selection of only using 500/404. This could also solve backward compatibility with currently only throwing 500.

I'll need to think a day or two.

**Comment by mosbth** [comment]
So I made som reading up on the subject and came up with the following.

First what status codes to use.

| Status codes | Use or no use
|------------------|-----------------------
| 400 Bad Request | Low level protocol issues (no use)
| 401 Unauthorized | Relates to WWW-Authenticate header ( no use)
| 403 Forbidden | I understand the request but refuse to serve it (use)
| 404 Not Found | No match is found, or refuse to serve and ignoring the reason (use)
| 500 Internal Server Error | Unexpected condition and failed to reply (use)

Sources: [HTTP/1.1 specification rfc2616](https://www.ietf.org/rfc/rfc2616.txt) and [StackOverflow on What HTTP status response code should I use if the request is missing a required parameter?](http://stackoverflow.com/questions/3050518/what-http-status-response-code-should-i-use-if-the-request-is-missing-a-required)

So, we can use 403, 404 and 500.

I propose the following use (in development and production mode).

| Status code proposed | Message or what happened |
|:---------------:|-------------------------------------|
| 500 | All exceptions thrown by `CImage`.
| 500 | Extension gd is not loaded.
| 500 | Unknown mode: $mode (relates to development, production, strict mode)
| 403 | Password required and does not match or exists.
| 403 | Hotlinking/leeching not allowed when password missmatch.
| 403 | Hotlinking/leeching not allowed and referer is missing.
| 403 | Hotlinking/leeching not allowed by whitelist. Referer: $referer.
| 403 | Hotlinking/leeching not allowed.
| 404 | Must set src-attribute.
| 404 | Filename contains invalid characters.
| 404 | Source image is not a valid file, check the filename and that a matching file exists on the filesystem.
| 404 | Security constraint: Source image is not below the directory "image_path" as specified in the config file img_config.php.
| 404 | Width % not numeric.
| 404 | Width out of range.
| 404 | Height % out of range.
| 404 | Height out of range.
| 404 | Aspect ratio out of range
| 404 | Quality out of range
| 404 | Compress out of range
| 404 | Scale out of range
| 404 | RotateBefore out of range
| 404 | RotateBefore out of range
| 403 | Directory for alias is not writable.
| 404 | Filename for alias contains invalid characters. Do not add extension.
| 403 | Alias is not enabled in the config file or password not matching.

To begin with, strict mode will use the same status codes as above. I was thinking about using only 404 for strict mode, but I leave that for now.

For the time being, no configuration in the config file, to affect the status codes returned.

I guess there is room for debate but this feels okey with me. Some of the 403 could be 404 instead, for example.

@Anatolie would you like to update your pull request to reflect above?

**Comment by Anatolie** [comment]
@mosbth I'll take a look at your proposed status codes and implement them in the next coming days in my branch.

I agree with having strict mode serving only 404's, but that can easily be added at a later date in the error() function whenever you decide to do so.

**Comment by mosbth** [comment]
Nice, looking forward to see it.

**Comment by Anatolie** [comment]
@mosbth Pushed out a commit with proposed changes in my pull request.

Now using just 403, 404 and 500 for the time being, like you proposed a few days ago.

Added changes to img.php, however I was not able to get the bash script to properly generate img[d/p/s]. For the time being, I reverted img[d/p/s] to their originals from the current version of the master branch.

**Could you please run the bash script on the updated img.php to generate img[d/p/s] files?**

**Comment by mosbth** [comment]
Sure thing, I'll merge it, test and package it.

**Comment by mosbth** [comment]
Merged into master and v0.7.8.

Thanks @Anatolie!
